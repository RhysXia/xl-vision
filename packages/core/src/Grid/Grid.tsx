import clsx from 'clsx'
import PropTypes from 'prop-types'
import React from 'react'
import { ColProps } from './Col'
import useMedia, { BreakPoint, breakPointArray } from './useMedia'
import RowContext from './GridContext'
import createUseStyles from '../styles/createUseStyles'

export type GridClassesKey =
  | 'root'
  | 'flex'
  | 'align-top'
  | 'align-middle'
  | 'align-bottom'
  | 'justify-start'
  | 'justify-end'
  | 'justify-center'
  | 'justify-space-around'
  | 'justify-space-between'

export type Align = 'top' | 'middle' | 'bottom'
export type Justify = 'start' | 'end' | 'center' | 'space-around' | 'space-between'

export interface GridProps extends React.HTMLAttributes<HTMLDivElement> {
  align?: Align
  children: React.ReactElement<ColProps> | Array<React.ReactElement<ColProps>>
  className?: string
  gutter?: number | Partial<Record<BreakPoint, number>>
  justify?: Justify
  type?: 'flex'
  classes?: Partial<Record<GridClassesKey, string>>
}

const Grid: React.FunctionComponent<GridProps> = (props) => {
  const { type, justify, align, className, style, children, gutter, classes, ...others } = props

  const builtinStyles = useStyles(classes)

  const media = useMedia()

  const computedGutter = React.useMemo(() => {
    if (typeof gutter === 'number') {
      return gutter
    }
    if (typeof gutter === 'object') {
      for (let i = 0; i < breakPointArray.length; i++) {
        const breakPoint = breakPointArray[i]
        if (media[breakPoint] && gutter[breakPoint] !== undefined) {
          return gutter[breakPoint] as number
        }
      }
    }
    return 0
  }, [media, gutter])

  const rootClassName = clsx(
    type === 'flex' ? builtinStyles.flex : builtinStyles.root,
    type === 'flex' && align && builtinStyles[`align-${align}` as 'align-top'],
    type === 'flex' && justify && builtinStyles[`justify-${justify}` as 'justify-start'],
    className
  )
  const rowStyle =
    computedGutter > 0
      ? {
          marginLeft: computedGutter / -2,
          marginRight: computedGutter / -2,
          ...style
        }
      : style

  return (
    <RowContext.Provider value={{ gutter: computedGutter, media }}>
      <div role='row' {...others} className={rootClassName} style={rowStyle}>
        {children}
      </div>
    </RowContext.Provider>
  )
}

Grid.displayName = 'Row'

Grid.propTypes = {
  align: PropTypes.oneOf<'top' | 'middle' | 'bottom'>(['top', 'middle', 'bottom']),
  children: PropTypes.oneOfType([
    PropTypes.element.isRequired,
    PropTypes.arrayOf(PropTypes.element.isRequired)
  ]).isRequired,
  className: PropTypes.string,
  gutter: PropTypes.oneOfType([PropTypes.number, PropTypes.object]),
  justify: PropTypes.oneOf<'start' | 'end' | 'center' | 'space-around' | 'space-between'>([
    'start',
    'end',
    'center',
    'space-around',
    'space-between'
  ]),
  type: PropTypes.oneOf<'flex'>(['flex']),
  style: PropTypes.object,
  classes: PropTypes.object
}
export default Grid

const useStyles = createUseStyles((theme) => {
  return {
    root: {
      position: 'relative',
      boxSizing: 'border-box',
      ...theme.mixins.clearfix
    },

    flex: {
      display: 'flex',
      flexDirection: 'row',
      '&:after': {
        display: 'none'
      }
    },
    'justify-start': {
      justifyContent: 'start'
    },
    'justify-end': {
      justifyContent: 'end'
    },
    'justify-center': {
      justifyContent: 'center'
    },
    'justify-space-around': {
      justifyContent: 'space-around'
    },
    'justify-space-between': {
      justifyContent: 'space-between'
    },
    'align-top': {
      alignItems: 'flex-start'
    },
    'align-middle': {
      alignItems: 'center'
    },
    'align-bottom': {
      alignItems: 'flex-end'
    }
  }
}, 'Row')
