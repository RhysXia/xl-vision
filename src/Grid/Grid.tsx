import classnames from 'classnames'
import PropTypes from 'prop-types'
import React from 'react'
import { ColProps } from './Col'
import useMedia, { BreakPoint, breakPointArray } from './useMedia'
import RowContext from './GridContext'
import ConfigContext from '../ConfigProvider/ConfigContext'

export interface GridProps extends React.HTMLAttributes<HTMLDivElement> {
  align?: 'top' | 'middle' | 'bottom'
  children: React.ReactElement<ColProps> | Array<React.ReactElement<ColProps>>
  className?: string
  gutter?: number | Partial<Record<BreakPoint, number>>
  justify?: 'start' | 'end' | 'center' | 'space-around' | 'space-between'
  clsPrefix?: string
  type?: 'flex'
}

const Grid: React.FunctionComponent<GridProps> = (props) => {
  const { clsPrefix: rootClsPrefix } = React.useContext(ConfigContext)

  const {
    type,
    justify,
    align,
    className,
    style,
    children,
    gutter,
    clsPrefix = `${rootClsPrefix}-grid`,
    ...others
  } = props

  const media = useMedia()

  const computedGutter = React.useMemo<number>(() => {
    if (typeof gutter === 'number') {
      return gutter
    }
    if (typeof gutter === 'object') {
      for (let i = 0; i < breakPointArray.length; i ++) {
        const breakPoint = breakPointArray[i]
        if (media[breakPoint] && gutter[breakPoint] !== undefined) {
          return gutter[breakPoint] as number
        }
      }
    }
    return 0
  }, [media, gutter])

  const classes = classnames(
    {
      [clsPrefix]: !type,
      [`${clsPrefix}-${type!}`]: type,
      [`${clsPrefix}-${type!}--${justify!}`]: type && justify,
      [`${clsPrefix}-${type!}--${align!}`]: type && align
    },
    className
  )
  const rowStyle =
    computedGutter > 0
      ? {
          marginLeft: computedGutter / -2,
          marginRight: computedGutter / -2,
          ...style
        }
      : style

  return (
    <RowContext.Provider value={{ gutter: computedGutter, media }}>
      <div {...others} className={classes} style={rowStyle}>
        {children}
      </div>
    </RowContext.Provider>
  )
}

Grid.displayName = 'Grid'

Grid.propTypes = {
  align: PropTypes.oneOf<'top' | 'middle' | 'bottom'>(['top', 'middle', 'bottom']),
  children: PropTypes.oneOfType([
    PropTypes.element.isRequired,
    PropTypes.arrayOf(PropTypes.element.isRequired)
  ]).isRequired,
  className: PropTypes.string,
  gutter: PropTypes.oneOfType([PropTypes.number, PropTypes.object]),
  justify: PropTypes.oneOf<'start' | 'end' | 'center' | 'space-around' | 'space-between'>([
    'start',
    'end',
    'center',
    'space-around',
    'space-between'
  ]),
  clsPrefix: PropTypes.string,
  type: PropTypes.oneOf<'flex'>(['flex']),
  style: PropTypes.object
}
export default Grid
