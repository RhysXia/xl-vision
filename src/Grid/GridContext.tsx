import React from 'react'
import { BreakPoint } from './useMedia'

export interface RowContextState {
  gutter: number
  media: Partial<Record<BreakPoint, boolean>>
}

const RowContext = React.createContext<RowContextState>({
  gutter: 0,
  media: {}
})

export default RowContext
