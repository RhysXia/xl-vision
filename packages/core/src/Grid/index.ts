import Col from './Col'
import Grid from './Grid'

export { BreakPoint } from './useMedia'
export { default as Col, ColProps, ColSpanType } from './Col'
export { default as Grid, GridProps } from './Grid'

const GridWithCol = Grid as typeof Grid & {
  Col: typeof Col
}

GridWithCol.Col = Col

export default GridWithCol
