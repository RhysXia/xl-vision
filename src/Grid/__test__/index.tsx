import { render } from 'enzyme'
import * as React from 'react'
import Grid from '..'

describe('Grid', () => {
  it('基本用法', () => {
    const wrapper = render(
      <div>
        <div className='box'>
          <Grid gutter={10}>
            <Grid.Col span={4}>
              <div className='col'>col1</div>
            </Grid.Col>
            <Grid.Col span={5}>
              <div className='col'>col2</div>
            </Grid.Col>
            <Grid.Col span={7}>
              <div className='col'>col3</div>
            </Grid.Col>
            <Grid.Col span={8}>
              <div className='col'>col4</div>
            </Grid.Col>
          </Grid>
        </div>
        <div className='box'>
          <Grid gutter={10}>
            <Grid.Col span={6}>
              <div className='col'>col1</div>
            </Grid.Col>
            <Grid.Col span={6}>
              <div className='col'>col2</div>
            </Grid.Col>
            <Grid.Col span={6}>
              <div className='col'>col3</div>
            </Grid.Col>
            <Grid.Col span={6}>
              <div className='col'>col4</div>
            </Grid.Col>
          </Grid>
        </div>
        <div className='box'>
          <Grid gutter={10}>
            <Grid.Col span={8}>
              <div className='col'>col1</div>
            </Grid.Col>
            <Grid.Col span={0}>
              <div className='col'>col2</div>
            </Grid.Col>
            <Grid.Col span={8}>
              <div className='col'>col3</div>
            </Grid.Col>
            <Grid.Col span={8}>
              <div className='col'>col4</div>
            </Grid.Col>
          </Grid>
        </div>
      </div>
    )
    expect(wrapper).toMatchSnapshot()
    // wrapper.unmount()
  })
  it('偏移布局', () => {
    const wrapper = render(
      <div>
        <div className='box'>
          <Grid gutter={10}>
            <Grid.Col span={6}>
              <div className='col'>col1</div>
            </Grid.Col>
            <Grid.Col span={6} offset={6}>
              <div className='col'>col2</div>
            </Grid.Col>
            <Grid.Col span={6}>
              <div className='col'>col3</div>
            </Grid.Col>
          </Grid>
        </div>
        <div className='box'>
          <Grid gutter={10}>
            <Grid.Col span={6}>
              <div className='col'>col1</div>
            </Grid.Col>
            <Grid.Col span={6} push={6}>
              <div className='col'>col2</div>
            </Grid.Col>
            <Grid.Col span={6} pull={6}>
              <div className='col'>col3</div>
            </Grid.Col>
            <Grid.Col span={6}>
              <div className='col'>col4</div>
            </Grid.Col>
          </Grid>
        </div>
        <div className='box'>
          <Grid gutter={10}>
            <Grid.Col span={6} offset={6}>
              <div className='col'>col1</div>
            </Grid.Col>
            <Grid.Col span={6} pull={12}>
              <div className='col'>col2</div>
            </Grid.Col>
            <Grid.Col span={6}>
              <div className='col'>col4</div>
            </Grid.Col>
          </Grid>
        </div>
      </div>
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('响应式布局', () => {
    const wrapper = render(
      <div className='box'>
        <Grid gutter={{ xs: 8, sm: 10, md: 15, lg: 20, xl: 25, xxl: 30 }}>
          <Grid.Col span={{ xs: 8, sm: 6, md: 4, lg: 6 }}>
            <div className='col'>col1</div>
          </Grid.Col>
          <Grid.Col span={{ xs: 8, sm: 6, md: 5, lg: 10, xxl: 6 }}>
            <div className='col'>col2</div>
          </Grid.Col>
          <Grid.Col span={{ xs: 8, sm: 6, md: 7, lg: 4, xl: 0, xxl: 6 }}>
            <div className='col'>col3</div>
          </Grid.Col>
          <Grid.Col span={{ xs: 0, sm: 6, md: 8, lg: 4, xl: 8, xxl: 6 }}>
            <div className='col'>col4</div>
          </Grid.Col>
        </Grid>
      </div>
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('响应式偏移', () => {
    const wrapper = render(
      <div>
        <div className='box'>
          <Grid gutter={10}>
            <Grid.Col span={6}>
              <div className='col'>col1</div>
            </Grid.Col>
            <Grid.Col span={6} offset={{ xs: 1, sm: 2, md: 3, lg: 4, xl: 5, xxl: 6 }}>
              <div className='col'>col2</div>
            </Grid.Col>
            <Grid.Col span={6}>
              <div className='col'>col3</div>
            </Grid.Col>
          </Grid>
        </div>
        <div className='box'>
          <Grid gutter={10}>
            <Grid.Col span={6}>
              <div className='col'>col1</div>
            </Grid.Col>
            <Grid.Col span={6}>
              <div className='col'>col2</div>
            </Grid.Col>
            <Grid.Col span={6} push={{ xs: 1, sm: 2, md: 3, lg: 4, xl: 5, xxl: 6 }}>
              <div className='col'>col3</div>
            </Grid.Col>
          </Grid>
        </div>
        <div className='box'>
          <Grid gutter={10}>
            <Grid.Col span={6}>
              <div className='col'>col1</div>
            </Grid.Col>
            <Grid.Col span={6}>
              <div className='col'>col2</div>
            </Grid.Col>
            <Grid.Col span={6} pull={{ xs: 1, sm: 2, md: 3, lg: 4, xl: 5, xxl: 6 }} offset={6}>
              <div className='col'>col3</div>
            </Grid.Col>
          </Grid>
        </div>
      </div>
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('flex布局', () => {
    const wrapper = render(
      <div>
        <div className='box'>
          <Grid gutter={10} type={'flex'} align={'top'}>
            <Grid.Col span={6}>
              <div className='col'>col1</div>
            </Grid.Col>
            <Grid.Col span={6}>
              <div className='col'>col2</div>
            </Grid.Col>
            <Grid.Col span={6}>
              <div className='col'>col3</div>
            </Grid.Col>
            <Grid.Col span={6}>
              <div className='col'>col4</div>
            </Grid.Col>
          </Grid>
        </div>
        <div className='box'>
          <Grid gutter={10} type={'flex'} align={'middle'}>
            <Grid.Col span={6}>
              <div className='col'>col1</div>
            </Grid.Col>
            <Grid.Col span={6}>
              <div className='col'>col2</div>
            </Grid.Col>
            <Grid.Col span={6}>
              <div className='col'>col3</div>
            </Grid.Col>
            <Grid.Col span={6}>
              <div className='col'>col4</div>
            </Grid.Col>
          </Grid>
        </div>
        <div className='box'>
          <Grid gutter={10} type={'flex'} align={'bottom'}>
            <Grid.Col span={6}>
              <div className='col'>col1</div>
            </Grid.Col>
            <Grid.Col span={6}>
              <div className='col'>col2</div>
            </Grid.Col>
            <Grid.Col span={6}>
              <div className='col'>col3</div>
            </Grid.Col>
            <Grid.Col span={6}>
              <div className='col'>col4</div>
            </Grid.Col>
          </Grid>
        </div>
        <div className='box'>
          <Grid gutter={10} type={'flex'} justify={'start'}>
            <Grid.Col span={6}>
              <div className='col'>col1</div>
            </Grid.Col>
            <Grid.Col span={6}>
              <div className='col'>col2</div>
            </Grid.Col>
            <Grid.Col span={6}>
              <div className='col'>col3</div>
            </Grid.Col>
          </Grid>
        </div>
        <div className='box'>
          <Grid gutter={10} type={'flex'} justify={'center'}>
            <Grid.Col span={6}>
              <div className='col'>col1</div>
            </Grid.Col>
            <Grid.Col span={6}>
              <div className='col'>col2</div>
            </Grid.Col>
            <Grid.Col span={6}>
              <div className='col'>col3</div>
            </Grid.Col>
          </Grid>
        </div>
        <div className='box'>
          <Grid gutter={10} type={'flex'} justify={'end'}>
            <Grid.Col span={6}>
              <div className='col'>col1</div>
            </Grid.Col>
            <Grid.Col span={6}>
              <div className='col'>col2</div>
            </Grid.Col>
            <Grid.Col span={6}>
              <div className='col'>col3</div>
            </Grid.Col>
          </Grid>
        </div>
        <div className='box'>
          <Grid gutter={10} type={'flex'}>
            <Grid.Col span={6} order={3}>
              <div className='col'>col1</div>
            </Grid.Col>
            <Grid.Col span={6} order={1}>
              <div className='col'>col2</div>
            </Grid.Col>
            <Grid.Col span={6} order={2}>
              <div className='col'>col3</div>
            </Grid.Col>
          </Grid>
        </div>
      </div>
    )
    expect(wrapper).toMatchSnapshot()
  })
})
