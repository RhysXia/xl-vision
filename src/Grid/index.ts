import Col from './Col'
import Grid from './Grid'

export { BreakPoint } from './useMedia'
export { ColProps, ColSpanType } from './Col'
export { GridProps } from './Grid'

const GridWithCol = Grid as typeof Grid & {
  Col: typeof Col
}

GridWithCol.Col = Col

export default GridWithCol
