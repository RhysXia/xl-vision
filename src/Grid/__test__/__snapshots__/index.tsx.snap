// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid flex布局 1`] = `
<div>
  <div
    class="box"
  >
    <div
      class="xl-grid-flex xl-grid-flex--top"
      style="margin-left:-5px;margin-right:-5px"
    >
      <div
        class="xl-col xl-col-span-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col1
        </div>
      </div>
      <div
        class="xl-col xl-col-span-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col2
        </div>
      </div>
      <div
        class="xl-col xl-col-span-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col3
        </div>
      </div>
      <div
        class="xl-col xl-col-span-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col4
        </div>
      </div>
    </div>
  </div>
  <div
    class="box"
  >
    <div
      class="xl-grid-flex xl-grid-flex--middle"
      style="margin-left:-5px;margin-right:-5px"
    >
      <div
        class="xl-col xl-col-span-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col1
        </div>
      </div>
      <div
        class="xl-col xl-col-span-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col2
        </div>
      </div>
      <div
        class="xl-col xl-col-span-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col3
        </div>
      </div>
      <div
        class="xl-col xl-col-span-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col4
        </div>
      </div>
    </div>
  </div>
  <div
    class="box"
  >
    <div
      class="xl-grid-flex xl-grid-flex--bottom"
      style="margin-left:-5px;margin-right:-5px"
    >
      <div
        class="xl-col xl-col-span-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col1
        </div>
      </div>
      <div
        class="xl-col xl-col-span-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col2
        </div>
      </div>
      <div
        class="xl-col xl-col-span-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col3
        </div>
      </div>
      <div
        class="xl-col xl-col-span-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col4
        </div>
      </div>
    </div>
  </div>
  <div
    class="box"
  >
    <div
      class="xl-grid-flex xl-grid-flex--start"
      style="margin-left:-5px;margin-right:-5px"
    >
      <div
        class="xl-col xl-col-span-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col1
        </div>
      </div>
      <div
        class="xl-col xl-col-span-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col2
        </div>
      </div>
      <div
        class="xl-col xl-col-span-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col3
        </div>
      </div>
    </div>
  </div>
  <div
    class="box"
  >
    <div
      class="xl-grid-flex xl-grid-flex--center"
      style="margin-left:-5px;margin-right:-5px"
    >
      <div
        class="xl-col xl-col-span-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col1
        </div>
      </div>
      <div
        class="xl-col xl-col-span-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col2
        </div>
      </div>
      <div
        class="xl-col xl-col-span-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col3
        </div>
      </div>
    </div>
  </div>
  <div
    class="box"
  >
    <div
      class="xl-grid-flex xl-grid-flex--end"
      style="margin-left:-5px;margin-right:-5px"
    >
      <div
        class="xl-col xl-col-span-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col1
        </div>
      </div>
      <div
        class="xl-col xl-col-span-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col2
        </div>
      </div>
      <div
        class="xl-col xl-col-span-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col3
        </div>
      </div>
    </div>
  </div>
  <div
    class="box"
  >
    <div
      class="xl-grid-flex"
      style="margin-left:-5px;margin-right:-5px"
    >
      <div
        class="xl-col xl-col-order-3 xl-col-span-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col1
        </div>
      </div>
      <div
        class="xl-col xl-col-order-1 xl-col-span-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col2
        </div>
      </div>
      <div
        class="xl-col xl-col-order-2 xl-col-span-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col3
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Grid 偏移布局 1`] = `
<div>
  <div
    class="box"
  >
    <div
      class="xl-grid"
      style="margin-left:-5px;margin-right:-5px"
    >
      <div
        class="xl-col xl-col-span-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col1
        </div>
      </div>
      <div
        class="xl-col xl-col-offset-6 xl-col-span-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col2
        </div>
      </div>
      <div
        class="xl-col xl-col-span-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col3
        </div>
      </div>
    </div>
  </div>
  <div
    class="box"
  >
    <div
      class="xl-grid"
      style="margin-left:-5px;margin-right:-5px"
    >
      <div
        class="xl-col xl-col-span-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col1
        </div>
      </div>
      <div
        class="xl-col xl-col-push-6 xl-col-span-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col2
        </div>
      </div>
      <div
        class="xl-col xl-col-pull-6 xl-col-span-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col3
        </div>
      </div>
      <div
        class="xl-col xl-col-span-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col4
        </div>
      </div>
    </div>
  </div>
  <div
    class="box"
  >
    <div
      class="xl-grid"
      style="margin-left:-5px;margin-right:-5px"
    >
      <div
        class="xl-col xl-col-offset-6 xl-col-span-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col1
        </div>
      </div>
      <div
        class="xl-col xl-col-pull-12 xl-col-span-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col2
        </div>
      </div>
      <div
        class="xl-col xl-col-span-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col4
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Grid 响应式偏移 1`] = `
<div>
  <div
    class="box"
  >
    <div
      class="xl-grid"
      style="margin-left:-5px;margin-right:-5px"
    >
      <div
        class="xl-col xl-col-span-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col1
        </div>
      </div>
      <div
        class="xl-col xl-col-span-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col2
        </div>
      </div>
      <div
        class="xl-col xl-col-span-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col3
        </div>
      </div>
    </div>
  </div>
  <div
    class="box"
  >
    <div
      class="xl-grid"
      style="margin-left:-5px;margin-right:-5px"
    >
      <div
        class="xl-col xl-col-span-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col1
        </div>
      </div>
      <div
        class="xl-col xl-col-span-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col2
        </div>
      </div>
      <div
        class="xl-col xl-col-span-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col3
        </div>
      </div>
    </div>
  </div>
  <div
    class="box"
  >
    <div
      class="xl-grid"
      style="margin-left:-5px;margin-right:-5px"
    >
      <div
        class="xl-col xl-col-span-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col1
        </div>
      </div>
      <div
        class="xl-col xl-col-span-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col2
        </div>
      </div>
      <div
        class="xl-col xl-col-offset-6 xl-col-span-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col3
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Grid 响应式布局 1`] = `
<div
  class="box"
>
  <div
    class="xl-grid"
  >
    <div
      class="xl-col"
    >
      <div
        class="col"
      >
        col1
      </div>
    </div>
    <div
      class="xl-col"
    >
      <div
        class="col"
      >
        col2
      </div>
    </div>
    <div
      class="xl-col"
    >
      <div
        class="col"
      >
        col3
      </div>
    </div>
    <div
      class="xl-col"
    >
      <div
        class="col"
      >
        col4
      </div>
    </div>
  </div>
</div>
`;

exports[`Grid 基本用法 1`] = `
<div>
  <div
    class="box"
  >
    <div
      class="xl-grid"
      style="margin-left:-5px;margin-right:-5px"
    >
      <div
        class="xl-col xl-col-span-4"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col1
        </div>
      </div>
      <div
        class="xl-col xl-col-span-5"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col2
        </div>
      </div>
      <div
        class="xl-col xl-col-span-7"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col3
        </div>
      </div>
      <div
        class="xl-col xl-col-span-8"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col4
        </div>
      </div>
    </div>
  </div>
  <div
    class="box"
  >
    <div
      class="xl-grid"
      style="margin-left:-5px;margin-right:-5px"
    >
      <div
        class="xl-col xl-col-span-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col1
        </div>
      </div>
      <div
        class="xl-col xl-col-span-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col2
        </div>
      </div>
      <div
        class="xl-col xl-col-span-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col3
        </div>
      </div>
      <div
        class="xl-col xl-col-span-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col4
        </div>
      </div>
    </div>
  </div>
  <div
    class="box"
  >
    <div
      class="xl-grid"
      style="margin-left:-5px;margin-right:-5px"
    >
      <div
        class="xl-col xl-col-span-8"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col1
        </div>
      </div>
      <div
        class="xl-col xl-col-span-0"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col2
        </div>
      </div>
      <div
        class="xl-col xl-col-span-8"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col3
        </div>
      </div>
      <div
        class="xl-col xl-col-span-8"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col4
        </div>
      </div>
    </div>
  </div>
</div>
`;
